### PR TITLE
Use grad_out for cudnn CTC loss

### LINKED ISSUE #26797: Inconsistent gradient from different backend of CTCLoss
## 🐛 Bug

I'm switching to the cudnn backend of CTCLoss since the other is not fully reproducible,
however, it turns out that the exact same model that used to work with pytorch's cuda backend ctc loss now failed.
With some simple example, I found that there's a huge difference in gradient (both direction and magnitude) between two backends.
I'm not sure if the bug is on pytorch or cudnn, but as far as I know, TensorFlow also used CTC from cudnn and there is no similar issue.
Thanks in advance.

## To Reproduce
```
import torch
torch.manual_seed(0)
torch.backends.cudnn.deterministic = True
torch.backends.cudnn.benchmark = False

batch_size = 16
seq_len = 50
target_len = seq_len//2
latent_dim = 128
vocab_size = 50

x = torch.randn((seq_len,batch_size,latent_dim)).cuda()
y = torch.randint(1,vocab_size,(batch_size*target_len,),dtype=torch.long).cuda()
x_len = seq_len*torch.ones((batch_size,),dtype=torch.long).cuda()
y_len = target_len*torch.ones((batch_size,),dtype=torch.long).cuda()
w = torch.nn.Linear(latent_dim,vocab_size).cuda()

def compute_ctc(x,y,x_len,y_len,use_cudnn):
    for p in w.parameters():
        if p.grad is not None:
            p.grad.zero_()
    # Forward
    output = w(x).log_softmax(dim=-1)
    if use_cudnn:
        loss = torch.nn.functional.ctc_loss(output,
                                            y.to('cpu',torch.int32),
                                            x_len.to('cpu',torch.int32),
                                            y_len.to('cpu',torch.int32))
    else:
        loss = torch.nn.functional.ctc_loss(output,y,x_len,y_len)
    # backward
    loss.backward()
    m, b = w.parameters()
    print('loss = {}\ngrad_norm = {}'.format(loss, m.grad.view(-1).norm()))
    return m.grad.clone()

print("===== Pytorch CTC =====")
torch_grad = compute_ctc(x,y,x_len,y_len,False)
print("===== Cudnn CTC =====")
cudnn_grad = compute_ctc(x,y,x_len,y_len,True)
print("===== Grad diff. =====")
print("Cos Sim. = ",torch.nn.functional.cosine_similarity(torch_grad.view(-1),cudnn_grad.view(-1),dim=0))
print("Magnitude  = ",cudnn_grad.view(-1).norm() / torch_grad.view(-1).norm())

```
## Behavior

```
===== Pytorch CTC =====
loss = 6.073373794555664
grad_norm = 0.4746553599834442
===== Cudnn CTC =====
loss = 6.073373794555664
grad_norm = 10431.4375
===== Grad diff. =====
Cos Sim. =  tensor(0.6508, device='cuda:0')
Magnitude  =  tensor(21976.8672, device='cuda:0')
```

## Environment

```
PyTorch version: 1.2.0
Is debug build: No
CUDA used to build PyTorch: 10.0.130

OS: Ubuntu 18.04.3 LTS
GCC version: (Ubuntu 7.4.0-1ubuntu1~18.04.1) 7.4.0
CMake version: Could not collect

Python version: 3.7
Is CUDA available: Yes
CUDA runtime version: 10.1.243
GPU models and configuration: 
GPU 0: GeForce GTX 1080 Ti
GPU 1: GeForce GTX 1080 Ti

Nvidia driver version: 430.50
cuDNN version: /usr/local/cuda-10.1/targets/x86_64-linux/lib/libcudnn.so.7

Versions of relevant libraries:
[pip3] numpy==1.17.2
[pip3] torch==1.2.0
[pip3] torchaudio==0.3.0
[pip3] torchvision==0.4.0
[conda] Could not collect

```

cc @ezyang @gchanan @zou3519

### PR DESCRIPTION
**Pull request by t-vi**
Using grad_out for CuDNN CTC loss fixes: #26797, #25833.

We also fix a cudnn incompatible change that surfaced during the testing: As of CuDNN 7.6 the semantics of the CTC loss gradients are different.
This leads us to disable CuDNN CTC for CuDNN < 7.6. To mitigate the impact on users, we convert the parameters for the native implementation if CuDNN isn't applicable (previously this would give an error.)

### CODE REVIEW

**Comment by t-vi** [comment]
The build failures are exactly my new test.
So to me, the tests look like the new backward (from derivatives.yaml) hasn't been picked up or something else is going wrong.
I cannot reproduce the failure locally with cuda10.1, I'll try cuda 9 next.

If you had a hint in addition, I'd also be very grateful.

**Comment by soumith** [comment]
> So to me, the tests look like the new backward (from derivatives.yaml) hasn't been picked up or something else is going wrong.

@ezyang ideas?

**Comment by ezyang** [comment]
> So to me, the tests look like the new backward (from derivatives.yaml) hasn't been picked up or something else is going wrong.

That's not my interpretation of the test failures. I see:

```
Sep 29 22:46:00 ======================================================================
Sep 29 22:46:00 FAIL: test_ctc_loss_cudnn_cpu (__main__.TestAutogradDeviceTypeCPU)
Sep 29 22:46:00 ----------------------------------------------------------------------
Sep 29 22:46:00 Traceback (most recent call last):
Sep 29 22:46:00   File "/var/lib/jenkins/workspace/test/common_device_type.py", line 161, in instantiated_test
Sep 29 22:46:00     return test(self, device_arg)
Sep 29 22:46:00   File "/var/lib/jenkins/workspace/test/common_device_type.py", line 298, in dep_fn
Sep 29 22:46:00     return fn(slf, device, *args, **kwargs)
Sep 29 22:46:00   File "test_autograd.py", line 3856, in test_ctc_loss_cudnn
Sep 29 22:46:00     self.assertEqual(grad_cudnn, grad_native, prec=1e-4)
Sep 29 22:46:00   File "/var/lib/jenkins/workspace/test/common_utils.py", line 698, in assertEqual
Sep 29 22:46:00     assertTensorsEqual(x, y)
Sep 29 22:46:00   File "/var/lib/jenkins/workspace/test/common_utils.py", line 668, in assertTensorsEqual
Sep 29 22:46:00     self.assertLessEqual(max_err, prec, message)
Sep 29 22:46:00 AssertionError: tensor(836.3644, device='cuda:0', dtype=torch.float32) not less than or equal to 0.0001 : 
```

If the derivatives.yaml change were not picked up, I'd expect to see some sort of shape error (due the unsqueezes not being applied). My guess is that it's cuDNN version related.

**Comment by t-vi** [comment]
Well I rebuilt locally with cuda 9 (but still cudnn 7.1.4) and it works. As far as I can tell it's cudnn 7.6 for the CI.  I will try upgrading cudnn here.

I'll try adding a backward to templates/Functions.cpp just to be sure (probably nicer, anyway, to not compute the gradient over and over in the ternary).

**Comment by t-vi** [comment]
So @ezyang was right: With CuDNN 7.1, the patch works well, with CuDNN 7.6, I get the same failure as the CI.

**Comment by ptrblck** [comment]
We are looking into it, as the test works fine in cudnn7500 and fails in cudnn7604.

**Comment by t-vi** [comment]
Unfortunately, I think this is still very broken.

File: aten/src/ATen/native/cudnn/LossCTC.cpp [hunk 1]
```diff
@@ -62,7 +62,13 @@ std::tuple<Tensor, Tensor> _cudnn_ctc_loss(const Tensor& log_probs_t, const Tens
 
   cudnnCTCLossAlgo_t algo = (deterministic ? CUDNN_CTC_LOSS_ALGO_DETERMINISTIC : CUDNN_CTC_LOSS_ALGO_NON_DETERMINISTIC);
 
-  Tensor probs = log_probs->softmax(2);
+  Tensor probs;
+  // before 7.6 passing probs was the only option, starting with 7.1 the log probs/unnormalized log probs is the default
```
Comment (ngimel) [inline aten/src/ATen/native/cudnn/LossCTC.cpp]: starting with 7.6? Nice find, it's not in the release notes unfortunately

File: test/test_autograd.py [hunk 2]
```diff
@@ -3832,6 +3832,29 @@ def ctc_after_softmax(x):
 
             gradcheck(ctc_after_softmax, [x], nondet_tol=1e-7)
 
+    @skipCUDAIfRocm
+    @unittest.skipIf(not (TEST_CUDNN and (TEST_CUDNN_VERSION if TEST_CUDNN_VERSION else 0) >= 7000), "needs cudnn >= 7.0")
```
Comment (ngimel) [inline test/test_autograd.py]: consider using device-generic cudnn decorators (added by #26791)

File: test/test_autograd.py [hunk 3]
```diff
@@ -3832,6 +3832,29 @@ def ctc_after_softmax(x):
...
+                                device='cuda', dtype=torch.long)
+        log_probs = torch.log_softmax(torch.randn(input_length, batch_size, num_labels, device='cuda', dtype=torch.float), 2)
+        log_probs.requires_grad_()
+
+        input_lengths = batch_size * [input_length]
+        target_lengths = batch_size * [target_length]
+        grad_out = torch.randn(batch_size, device='cuda', dtype=torch.float)
+        loss_native = torch.nn.functional.ctc_loss(log_probs, targets, input_lengths, target_lengths, reduction='none')
+        grad_native, = torch.autograd.grad(loss_native, log_probs, grad_out)
+        loss_cudnn = torch.nn.functional.ctc_loss(log_probs, targets.to('cpu', torch.int32),
```
Comment (ngimel) [inline test/test_autograd.py]: As I see, the thing that triggers cudnn use (compared to line 3850) is targets being on the cpu and of int32 type? That seems fragile and non-obvious in the test, can you put line 3850 in the context manager disabling cudnn?

**Review by ngimel — APPROVED** [review]
approved, subject to minor fixes in the test

File: aten/src/ATen/native/cudnn/LossCTC.cpp (hunk 1 above)
Comment (t-vi) [reply aten/src/ATen/native/cudnn/LossCTC.cpp]: Unfortunately, the fix isn't working, either, as one would need to set the non-default option or somesuch.

**Comment by t-vi** [comment]
So this is all pretty terrible. As an alternative approach, I tried to use the default input options for CuDNN 7.6 and using softmax outside the "core function" (so that the backward of softmax is used), I do get differences in the derivatives. These are smaller and only for some inputs, apparently, but given that the native function does seem to pass gradcheck in the snippet below.

```python
batch_size = 1
input_length = 10
num_labels = 3
target_length = 5
targets = torch.randint(1, num_labels, (batch_size * target_length,),
                        device='cuda', dtype=torch.long)
log_probs = torch.log_softmax(torch.randn(input_length, batch_size, num_labels, device='cuda', dtype=torch.float), 2)
log_probs.requires_grad_()

input_lengths = batch_size * [input_length]
target_lengths = batch_size * [target_length]
grad_out = torch.randn(batch_size, device='cuda', dtype=torch.float)
with torch.backends.cudnn.flags(enabled=False):
    loss_native = torch.nn.functional.ctc_loss(log_probs.double(), targets, input_lengths, target_lengths, reduction='none')
    grad_native = torch.autograd.grad(loss_native, log_probs, grad_out)[0].float()
torch.autograd.gradcheck(lambda x: torch.nn.functional.ctc_loss(x.log_softmax(2), targets, input_lengths, target_lengths, reduction='sum'), log_probs.double())
loss_cudnn = torch.nn.functional.ctc_loss(log_probs, targets.to('cpu', torch.int32),
                                                   input_lengths, target_lengths, reduction='none')
grad_cudnn, = torch.autograd.grad(loss_cudnn, log_probs, grad_out)
print((grad_cudnn-grad_native).max())
```

**Comment by t-vi** [comment]
Just a quick note: The obvious alternative is to just support CuDNN >= 7.6 and not hack around 7.1 compatibility issues.

**Comment by soumith** [comment]
please disable cudnn path for CuDNN < 7.6 and let's get this patch in

**Comment by t-vi** [comment]
This is still broken, apparently, e.g. `pytorch_namedtensor_linux_xenial_cuda9_cudnn7_py2_test`.

**Pull request closed without merging**